recipes/mimalloc/all/test_package/conanfile.py

```diff
@@ -0,0 +1,35 @@
+from conans import ConanFile, CMake, tools
+from conans.model.options import PackageOptions
+import os
+
+
+class MimallocTestConan(ConanFile):
+    settings = "os", "compiler", "build_type", "arch"
+    generators = "cmake"
+    _test_file = "test_basic"
+    _build_basic = False
+    _build_redefine_malloc = False
+
+    def configure(self):
+        options = PackageOptions(self.options["mimalloc"])
+        if options.get_safe("redefine_malloc"):
+            self._test_file = "test_redefine_malloc"
+            self._build_redefine_malloc = True
```

Comment:
Nice way to be able to use `get_safe` in the test_package.

@jgsogo 
Shouldn't the `self.options` in test_package always be a `PackageOptions` object? Because now, none of the attributes such as `get_safe` are available in a test_package.

Reply:
`self.options` in any recipe is the same kind of object (afaik), here I see a "·$%"·! workaround to get a `PackageOptions` object from `self.options["mimalloc"]`. **Please, use only documented classes/methods in recipes.**

Here, something like `if 'redefine_malloc' in self.options["mimalloc"] and self.options["mimalloc"].redefine_malloc:` will do the work.

OTH, I cannot see an option `redefine_malloc` declared in the recipe, where is it coming from?

Reply:
@jgsogo 
`self.options["mimamalloc"]` is a different object type as `self.options`.
It would be nice if the first also has similar behavior.

`self.options.get_safe` works.
`self.options["mimalloc"].get_safe` does NOT work.

Reply:
I thought that the above approach was working, but sadly no =/

Indeed, since I've changed things a bit the correct test would be something like this (after implementing @jgsogo 's review):

```python
def configure(self):
    single_object = "single_object" in self.options["mimalloc"] and \
                    self.options["mimalloc"].single_object
    if not single_object and \
            not self.options["mimalloc"].shared and \
            self.options["mimalloc"].override:
        self._test_file = "test_redefine_malloc"
        self._build_redefine_malloc = True
    else:
        self._build_basic = True
```

But that doesn't seem to take the recipe's default values into account. When I call Conan like this:

```bash
$ conan create -s compiler.cppstd=14 -o mimalloc:override=True . mimalloc/1.6.4@ --build=mimalloc
```
I get the following error:

```
ERROR: mimalloc/1.6.4 (test package): Error in configure() method, line 16
	not self.options["mimalloc"].shared and \
	ConanException: option 'shared' doesn't exist
Possible options are ['override']
```

even though the recipe doesn't delete the `shared` option and defaults it to `False`. It seems that it's only considering the options passed in the command line, for, if I issue the following:

```bash
$ conan create -s compiler.cppstd=14 -o mimalloc:override=True -o mimalloc:shared=False . mimalloc/1.6.4@ --build=mimalloc
```
everything runs smoothly.

Reply:
> @jgsogo
> `self.options["mimamalloc"]` is a different object type as `self.options`.
> It would be nice if the first also has similar behavior.
> 
> `self.options.get_safe` works.
> `self.options["mimalloc"].get_safe` does NOT work.

Ok, this is a feature request for Conan. I would expect the same interface (and documentation) for both objects. Will you open the issue, please?

Reply:
Issue reported by @tarc, yes, in the `configure` method the options from the requirements are not available. The graph is constructed from the root to the requirements branches taking into account values from options to add/remove requirements and modify the topology of the graph (that's the reason why options from the requirements are not available while building the graph, it would be a chicken-and-egg problem).

> If those options are passed from the command-line (or overridden by other packages) the values are available because they are propagated from the root of the graph (Conan 2.0 will be more consistent with this).

You need to move this kind of checks that retrieve the value from the requirements to the `build()` method, at that moment, the graph is fully computed and all options are resolved and available.

Reply:
Hey, @jgsogo . That's a fine piece of information I totally missed!

Reply:
> > @jgsogo
> > `self.options["mimamalloc"]` is a different object type as `self.options`.
> > It would be nice if the first also has similar behavior.
> > `self.options.get_safe` works.
> > `self.options["mimalloc"].get_safe` does NOT work.
> 
> Ok, this is a feature request for Conan. I would expect the same interface (and documentation) for both objects. Will you open the issue, please?

@jgsogo Opened as https://github.com/conan-io/conan/issues/7548
Apparently, i opened it as a bug.